plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/PomXmlValidator.java

```diff
@@ -16,35 +16,52 @@
 
 package com.google.cloud.tools.eclipse.appengine.validation;
 
-import java.io.IOException;
-import java.util.Map;
+import java.util.ArrayList;
 
-import javax.xml.parsers.ParserConfigurationException;
+import org.eclipse.core.resources.IResource;
+import org.w3c.dom.Document;
+import org.w3c.dom.Node;
+import org.w3c.dom.NodeList;
 
-import org.eclipse.core.resources.IFile;
-import org.eclipse.core.runtime.CoreException;
-import org.xml.sax.SAXException;
+import com.google.common.annotations.VisibleForTesting;
 
-public class PomXmlValidator extends AbstractXmlValidator {
+public class PomXmlValidator implements XmlValidationHelper {
 
-  /**
-   * Clears all problem markers from the resource, then adds a Maven plugin marker
-   * to pom.xml if the App Engine Maven plugin is obsolete.
-   */
-  @Override
-  protected void validate(IFile resource, byte[] bytes)
-      throws CoreException, IOException, ParserConfigurationException {
-    try {
-      deleteMarkers(resource);
-      SaxParserResults parserResults = PomParser.readXml(bytes);
-      Map<BannedElement, Integer> bannedElementOffsetMap =
-          ValidationUtils.getOffsetMap(bytes, parserResults);
-      for (Map.Entry<BannedElement, Integer> entry : bannedElementOffsetMap.entrySet()) {
-        createMarker(resource, entry.getKey(), entry.getValue());
+  @VisibleForTesting
+  public ArrayList<BannedElement> checkForElements(IResource resource, Document document) {
+    DocumentLocation location = null;
+    ArrayList<BannedElement> blacklist = new ArrayList<>();
+    NodeList nodeList = document.getElementsByTagName("plugin");
+    
+    for (int i = 0; i < nodeList.getLength(); i++) {
+      Node node = nodeList.item(i);
+      NodeList childNodes = node.getChildNodes();
+      boolean foundGroupId = false;
+      boolean foundArtifactId = false;
+      String nodeText = "";
+      
+      for (int j = 0; j < childNodes.getLength(); j++) {
+        Node pluginChild = childNodes.item(j);
+        String localName = pluginChild.getNodeName();
+        nodeText = pluginChild.getTextContent();
```

Comment:
You'll end up always adding the text of the last `childNodes`. If you want to add the `artifactId` element, I think this should move into the `if ("artifactId"...)` branch.